crates/nu-protocol/src/errors/cli_error.rs

```diff
@@ -20,6 +28,31 @@ struct CliError<'src>(
     pub &'src StateWorkingSet<'src>,
 );
 
+#[derive(Default)]
+struct ReportLog {
+    deprecation_warnings: HashSet<DeclId>,
+}
+
+/// Returns true if this warning should be reported
+fn should_show_warning(warning: &ParseWarning) -> bool {
+    if let Some(decl_id) = match warning {
+        ParseWarning::DeprecatedWarning { decl_id, .. } => Some(decl_id),
+        ParseWarning::DeprecatedWarningWithMessage { decl_id, .. } => Some(decl_id),
+        // this is unreachable right now, but future additions to ParseWarning shouldn't be matched here
+        #[allow(unreachable_patterns)]
+        _ => None,
+    } {
+        let is_first_report = REPORT_LOG
+            .lock()
+            .expect("report log is poisioned")
```

Comment:
In our dependecy graph is already `parking_lot`, it provides a Mutex which doesn't poison. Maybe that's better.

Reply:
no longer relevant after moving to engine state